### PR TITLE
Add Bucket Storage Support for files

### PR DESCRIPTION
**Pull request by masudur-rahman**
- Add bucket storage support for the following items:
	- `avatars`
	- `repo-avatars`
	- `attachments`
	- `lfs`

### CODE REVIEW

File: modules/setting/setting.go
Comment (lafriks) [inline modules/setting/setting.go]: This should not be changed

File: modules/setting/setting.go
Comment (lafriks) [inline modules/setting/setting.go:702]: This breaks current behavior and is not really related to this PR

**Review by lafriks — CHANGES_REQUESTED** [review]

File: modules/setting/setting.go
Comment (masudur-rahman) [reply modules/setting/setting.go]: @lafriks ,I need the followings:
```go
AvatarUploadPath = "data/avatars/"
LFS.ContentPath	 = "data/lfs/"
AttachmentPath	 = "data/attachments/"
RepositoryAvatarUploadPath = "data/repo-avatars/"
```

Should I use:
`LFS.ContentPath = sec.Key("LFS_CONTENT_PATH").MustString(filepath.Join("data", "lfs/"))`
instead of:
`LFS.ContentPath = sec.Key("LFS_CONTENT_PATH").MustString(filepath.Join(AppDataPath, "lfs"))`

Comment (lafriks) [reply modules/setting/setting.go]: Problem is that this is just defaults. User could have specified completely different paths in their app.ini for ex:
```ini
AvatarUploadPath = "/srv/avatars/upload/"
LFS.ContentPath	 = "/srv/lfs/"
AttachmentPath	 = "/srv/attachments/"
RepositoryAvatarUploadPath = "/srv/avatars/repo-upload/"
```

And in such configuration your code will not work for local storage anymore

Comment (masudur-rahman) [reply modules/setting/setting.go]: It will work for either local or any other cloud storage.
By default I set it to local storage. https://github.com/go-gitea/gitea/blob/16ac0bb98f889f15628e1779889fd0b689927f2a/modules/setting/file_storage.go#L10-L23
If user wants to use any other cloud storage or other local storage, they just have to specify it in their `app.ini`

Users are free to provide their own upload path like you suggested and it will work fine.

Comment (lafriks) [reply modules/setting/setting.go]: I meant it should work for local storage same as it was before without changing app.ini

Comment (masudur-rahman) [reply modules/setting/setting.go]: Yes, it will work as it was before. Just the files will be stored in the directory specified in `FileStorage.BucketURL`.

So, what do you suggest about https://github.com/go-gitea/gitea/pull/7795#discussion_r311939134 ?

**Comment by masudur-rahman** [comment]
Hey @lafriks , previously I misunderstood you.
I have changed the code to be backward compatible. Please, let me know if anything needs to be changed.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md
Comment (lafriks) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: This change does not seems to be related to this PR

File: models/migrations/migrations.go
Comment (lafriks) [inline models/migrations/migrations.go]: Why checking only for absolute attachment path?

File: models/migrations/v61.go
Comment (lafriks) [inline models/migrations/v61.go]: IMHO it should be checking for some other setting not if path is absolute

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md
Comment (masudur-rahman) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md]: It's just a new line at the end.

File: models/migrations/migrations.go
Comment (masudur-rahman) [reply models/migrations/migrations.go]:
```
AttachmentPath	 = "/srv/attachments/"
```
If the AttachmentPath is absolute, prefix "file://" must be added to it in order to open it using `blob.OpenBucket`.
That's why I checked it.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md.orig
Comment (lunny) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md.orig]: This file should not be added to git.

File: models/admin.go
Comment (lunny) [inline models/admin.go]: Wrong imports order.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
I think we should have an abstract layout under `modules/` i.e. `modules/fs` to let people chose file system or block storage.

File: custom/conf/app.ini.sample
Comment (guillep2k) [inline custom/conf/app.ini.sample]: "`backward compatibility`" makes it sound like buckets are the "modern" way to go. Could this be changed to something like "`if unset, {APP_DATA_PATH}/attachments is used`" (not checked) or something around those lines?

File: models/admin.go
Comment (guillep2k) [inline models/admin.go]: Is there a way to make an error here gracefully fail, so the user can attempt to retry the operation? Third-party requests are likely to fail at some point.

File: models/attachment.go
Comment (guillep2k) [inline models/attachment.go]: Same thing as above about gracefully return connection errors.

File: custom/conf/app.ini.sample
Comment (masudur-rahman) [reply custom/conf/app.ini.sample]: Actually if unset, it now uses `AppWorkPath`. When using this path for `attachments` or `avatars`, `AttachmentPath` or `AvatarUploadPath` is appended to this `BucketURL`.
If `AttachmentPath` is unset, we set this as `data/attachments`, to maintain similarity between all the available buckets. https://github.com/go-gitea/gitea/blob/83ccf4c649e58abb83c7d21ec42a3f378be51949/modules/setting/setting.go#L803

Comment (guillep2k) [reply custom/conf/app.ini.sample]: So, if I understand well, the user can't use `APP_DATA_PATH` anymore? I think it's too big of a breaking change. If the user upgrades because they were interested in X other functionality the new version provides, he or she might not be aware of this change and get in a lot of trouble. Can't they both co-exist? e.g. if there are both `APP_DATA_PATH` and `BUCKET_URL`, then the former should be ignored, but if there's only `APP_DATA_PATH`, the old behaviour should be honored.

Comment (masudur-rahman) [reply custom/conf/app.ini.sample]: `APP_DATA_PATH` related issue is now fixed. 
https://github.com/go-gitea/gitea/blob/4c390c9699ee6400a9467c4c8234af9efba5acaf/modules/setting/setting.go#L1053-L1083

Now, `APP_DATA_PATH`and `BUCKET_URL` both co-exist. Please let me know if I did it right.

File: models/admin.go
Comment (masudur-rahman) [reply models/admin.go]: The error related to `OpenBucket` is handled by the bucket providers using `exponential backoff` algo. I think it's okay not to handle it here.

**Comment by masudur-rahman** [comment]
> I think we should have an abstract layout under `modules/` i.e. `modules/fs` to let people chose file system or block storage.

Hey @lunny, I am working on that now. I will make those changes in a new pr. If the changes look good, we can copy the changes in this pr later. 
Please take a look if the changes are okay or any thing needs to be changed. Thanks

File: models/migrations/migrations.go
Comment (masudur-rahman) [reply models/migrations/migrations.go]: I have updated it. I've attached a summary how whole `BucketURL` and `AttachmentPath` or others work.
https://github.com/go-gitea/gitea/blob/4c390c9699ee6400a9467c4c8234af9efba5acaf/modules/setting/setting.go#L1053-L1083

Please take a look if it seems okay. Thanks

File: modules/setting/setting.go
Comment (guillep2k) [reply modules/setting/setting.go]: So I've checked out this PR and ran it with my usual test settings:

```
[server]
SSH_DOMAIN       = ....
DOMAIN           =  ....
HTTP_PORT        =  ....
PROTOCOL         = http
ROOT_URL         =  ....
OFFLINE_MODE     = false
APP_DATA_PATH    = /mnt/external/my_test_data
```

Then I've changed my avatar. I _expected_ the avatar to be created at:
```
/mnt/external/my_test_data/avatars/a9050bf53d1005dfe5398ce9fa5b07aa
```

Instead, with this PR, the avatar was **incorrectly** created at:
```
/home/giteatest/avatars/a9050bf53d1005dfe5398ce9fa5b07aa
```

As you can see, with this PR Gitea completely ignored the value of my `APP_DATA_PATH`. This is the breaking change we've been talking about. The code must behave **like it did before this PR** if the user doesn't change their `app.ini`. Users that upgrade to this must not notice any difference, unless they actively decide to use the new bucket storage capability.

I hope that this example helps.

Comment (masudur-rahman) [reply modules/setting/setting.go]: Hey @guillep2k, thanks for taking the time for testing.

If `AVATAR_UPLOAD_PATH` is set either relative or absolute, it ignores `APP_DATA_PATH` as it was ignored previously. https://github.com/go-gitea/gitea/blob/111d31d6810c6a1336db6897b8cf1b249f5d709a/modules/setting/setting.go#L850-L854

But if `AVATAR_UPLOAD_PATH` is not specified, then it uses `APP_DATA_PATH`+"avatars" to read or store the avatars.

Either way if the `AvatarUploadPath` is relative then the whole path to store the avatars is `AppWorkPath`+`AvatarUploadPath`.
If the path is absolute the avatars is stored to `AvatarUploadPath`.
It works like it did previously.

I've attached the scenarios here : https://github.com/go-gitea/gitea/blob/e29ba594b069ae3011601ba210b40cbe9a057ff1/modules/storage/storage.go#L124-L154

May be there was some errors in my previous commit but I refactored the code and I hope it works like it did before this pr.

**Please, pull again and check if it's now okay. Thanks again.**

**Comment by masudur-rahman** [comment]
@lunny , I have refactored this pr. I moved the `FileStorage` thing to `modules/storage`.
https://github.com/go-gitea/gitea/blob/e29ba594b069ae3011601ba210b40cbe9a057ff1/modules/storage/storage.go

I hope these changes serve the abstract layout you asked for. 
Please let me know if everything's okay. Thanks.

**Comment by tamalsaha** [comment]
@lunny / @lafriks / @guillep2k and folks, if you can take another look at it that will be great.

**Comment by guillep2k** [comment]
Now the `APP_DATA_PATH` setting seems to be working (at least for avatars), but for some reason uploading a picture leaves an `.attrs` file behind that I don't think I've seen before:

```
-rw-------+ 1 giteatest giteatest 269836 Aug 22 22:17 my_test_data/avatars/9-eae2c62356750433ffc3c52ffb9ceee9
-rw-rw----+ 1 giteatest giteatest    196 Aug 22 22:17 my_test_data/avatars/9-eae2c62356750433ffc3c52ffb9ceee9.attrs
```

Contents:
```
{"user.cache_control":"","user.content_disposition":"","user.content_encoding":"","user.content_language":"","user.content_type":"image/png","user.metadata":null,"md5":"a+USfqDVjIMq4Sy3HXOWdA=="}
```

Why is it there? Is that a side-effect of your changes?

**Comment by tamalsaha** [comment]
@guillep2k , this is a result of using the gocloud/blob/filestorage api. This library stores file attributes in this .attrs file .

- https://github.com/google/go-cloud/blob/master/blob/fileblob/fileblob.go
- https://github.com/google/go-cloud/pull/203

**Comment by tamalsaha** [comment]
Please take a look. The git conflicts are fixed.

**Comment by guillep2k** [comment]
Weird fail in the build. Somebody will restart it soon.

**Comment by lunny** [comment]
I will review this ASAP.

**Comment by masudur-rahman** [comment]
@lunny, I keep getting `build fail` after resolving the conflicts. I ran `go mod tidy && go mod vendor` every time after resolving the conflicts. 
But there is some dissimilarity in `vendor/modules.txt` between my version and `continuous-integration/drone/pr` version which I don't know why.

Can you please give me some suggestions what should I do to resolve this problem ?

**Comment by lunny** [comment]
@masudur-rahman CI failed because of other reason and just fixed. Please resolve the conflicts.

**Comment by masudur-rahman** [comment]
@lunny, I resolved the conflicts and kept the changes in a single commit. Still getting `build fail` due to `modules.txt`.

**Comment by masudur-rahman** [comment]
At last the `continuous-integration/drone/pr` is passing. May be it will be failing after merging some other pr. 
@lunny , @lafriks & @guillep2k , can you guys please take a look while `continuous-integration/drone/pr` is passing ?

Thanks...

File: models/attachment.go
Comment (lunny) [inline models/attachment.go]: just keep `defer fw.Close()` is better.

Comment (masudur-rahman) [reply models/attachment.go]: `defer fw.Close()` can't be used here.
`fw` must be closed before `fi, err := fs.Attributes()` (Line 105). It can't open the file for attributes unless the `fw.Close()` is called.

That's why I had to do like this.

File: models/repo.go
Comment (guillep2k) [inline models/repo.go:2808]: What happens if two users choose the same file name for their repo avatars?

File: models/user.go
Comment (guillep2k) [inline models/user.go:541]: Same question as for repo avatar file name.

File: modules/migrations/gitea.go
Comment (guillep2k) [inline modules/migrations/gitea.go:262]: This is trickier. Depeding on the current version value you should or should not ignore the app.ini setting and use local file storage in this function.

File: modules/public/avatars.go
Comment (guillep2k) [inline modules/public/avatars.go:19]: Is this necessary? The new logger system allows to set logging levels for different features individually.

File: modules/public/avatars.go
Comment (guillep2k) [inline modules/public/avatars.go:81]: I think users (admins) should decide which resources are handled locally or remotely. Avatars are too time-sensitive to keep in a remote server, especially if they need to pass through Gitea to be retrieved. Attachments are usually the images of issues and PRs, so they too are pretty much time-sensitive. The `Etag` of course helps a lot. Nonetheless, I'd have three separate settings: one for avatars, one for attachments and one for LFS. Even let the admin set different services for each one.

File: modules/setting/setting.go
Comment (guillep2k) [inline modules/setting/setting.go:800]: As said above, I'd have three sections: `[storage.avatars]`, `[storage.attachments]`, `[storage.lfs]`. Or some flexible system like that.

File: modules/storage/storage.go
Comment (guillep2k) [inline modules/storage/storage.go:29]: If you pass an enum here, it's easy to decide between different file storages.

**Review by guillep2k — COMMENTED** [review]
Decidedly _not_ an easy PR to review. 😅

**Comment by guillep2k** [comment]
One important aspect that I didn't see covered: how does one user migrate from one storage system to another? For example: I decide I need aws for my files, but my repository is already up and running. If I change `app.ini`, all my files will be suddenly gone! I'd need some tool to transfer all my files to the new storage and _then_ start Gitea.

File: models/attachment.go
Comment (masudur-rahman) [reply models/attachment.go]: Sorry for the late. 
Your suggestion will work fine here and will do that.

File: models/migrations/v61.go
Comment (masudur-rahman) [reply models/migrations/v61.go:39]: Yes, that's very necessary.

File: models/repo.go
Comment (masudur-rahman) [reply models/repo.go:2808]: Here, `<repo.ID>-<md5.Sum(data)>` is used as  `repo.Avatar`. So even if two user uses same file,  repo.ID will differ them.
https://github.com/go-gitea/gitea/blob/98c0c8ea2e0d4896cc3cc00423f0d2952fcbad21/models/repo.go#L2718

File: models/user.go
Comment (masudur-rahman) [reply models/user.go:541]: Same answer as for repo.avatar.
But previously, if two users uploaded the same file, it would replace the previous one (also in previous system).
https://github.com/go-gitea/gitea/blob/d4cd4ed4422be222088dd8535a228b906f43fdc2/models/user.go#L523

I used same format as `repo.Avatar` to avoid this collision.
https://github.com/go-gitea/gitea/blob/98c0c8ea2e0d4896cc3cc00423f0d2952fcbad21/models/user.go#L533

File: modules/public/avatars.go
Comment (masudur-rahman) [reply modules/public/avatars.go:19]: I followed the existing `Options` structure to handle this part. 
https://github.com/go-gitea/gitea/blob/d4cd4ed4422be222088dd8535a228b906f43fdc2/modules/public/public.go#L24-L33

File: modules/storage/storage.go
Comment (masudur-rahman) [reply modules/storage/storage.go:29]: If we want to separate storage system for avatars, lfs, attachments ( `[storage.avatars]`, `[storage.attachments]`, `[storage.lfs]` ), we  can use an extra field `BucketURL` here which will be useful to decide between file storages. 
https://github.com/go-gitea/gitea/blob/98c0c8ea2e0d4896cc3cc00423f0d2952fcbad21/modules/storage/storage.go#L22-L26

File: modules/public/avatars.go
Comment (guillep2k) [reply modules/public/avatars.go:19]: I didn't realize that.

**Comment by guillep2k** [comment]
A migration tool will be needed indeed. Between different cloud storages I don't know, but at least between any cloud storage and file://. And it should be reversible (e.g. file names should not be problem, even if e.g. avatars get a different name when migrated).

**Comment by lunny** [comment]
Should we provide a new command to copy local directory data to a storage?

**Comment by guillep2k** [comment]
> 
> 
> Should we provide a new command to copy local directory data to a storage?

@lunny You mean something like this?:

> 
> 
> A migration tool will be needed indeed. Between different cloud storages I don't know, but at least between any cloud storage and file://. And it should be reversible (e.g. file names should not be problem, even if e.g. avatars get a different name when migrated).

**Comment by lunny** [comment]
I would like we have an abstract filesystem layer before this PR.  Something like https://godoc.org/github.com/blang/vfs or other better libraries. So we could support both local file system, blob storages and other storage facilities if there are implements for them.

**Comment by lunny** [comment]
Or https://github.com/spf13/afero

**Comment by patcito** [comment]
Any news about this? @masudur-rahman

**Pull request closed without merging**
